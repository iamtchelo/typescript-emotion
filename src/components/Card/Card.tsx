import * as React from 'react';
import { CardImage, CardTitle, CardWrapper } from './card.elements';

export interface CardProps {
  title: string;
  image: string;
}

const Card: React.SFC<CardProps> = ({
  title,
  image
}) => (
  <CardWrapper>
    <CardImage src={image} />
    <CardTitle>{title}</CardTitle>
  </CardWrapper>
);

export default Card;