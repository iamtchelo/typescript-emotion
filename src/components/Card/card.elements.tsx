import styled from 'react-emotion';

export const CardWrapper = styled('div')`
  display: flex;
  width: 100%;
  flex-direction: column;
  justify-content: center;
  background: ${(props: any) => props.theme.colors.primary};
  box-shadow: 0 0 5px rgba(0, 0, 0, 0.1);
  transition: transform 0.2s ease-out;

  &:hover {
    background: ${(props: any) => props.theme.colors.primaryHover};
  }

  &:nth-child(even):hover {
    transform: scale(1.15) rotate(10deg);
  }

  &:nth-child(odd):hover {
    transform: scale(1.15) rotate(-10deg);
  }
`;

export const CardImage = styled('img')`
  margin: auto;
  padding: 5px;
  width: 100%;
  height: auto;
  border: 5px solid ${(props: any) => props.theme.colors.white};
`;

export const CardTitle = styled('h4')`
  padding: 10px 0;
  text-align: center;
  background: ${(props: any) => props.theme.colors.white};
  font-weight: normal;
  text-transform: capitalize;
  color: ${(props: any) => props.theme.card.cardTitle};
`;
