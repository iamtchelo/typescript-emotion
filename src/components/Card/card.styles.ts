import styled, {css} from 'react-emotion';

const cardBase = ({theme}: any) => css`
  display: flex;
  width: 100%;
  flex-direction: column;
  justify-content: center;
  background-color: ${theme.colors.primary};
  box-shadow: 0 0 5px rgba(0, 0, 5px, rgba(0, 0, 0, 0.1));

  &:hover {
    box-shadow: 0 0 10px rgba(0, 0, 0, 0.3);
    background-color: ${theme.colors.primaryHover};
  }
`;

const cardImage = () => css`
  margin: auto;
  padding: 5px;
  width: 100%;
  height: auto;
  border: 5px solid white;
`;

const cardTitle = ({theme}: any) => css`
  padding: 10px 0;
  text-align: center;
  text-transform: capitalize;
  font-weight: normal;
  background-color: white;
  color: ${theme.card.cardTitle};
`;

const CardWrapper = styled('div')`
  ${cardBase};
`;

const CardImage = styled('img')`
  ${cardImage};
`;

const CardTitle = styled('h4')`
  ${cardTitle};
`;

export {CardWrapper, CardImage, CardTitle};
