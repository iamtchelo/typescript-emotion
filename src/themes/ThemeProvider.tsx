import { ThemeProvider as Provider } from 'emotion-theming';
import * as React from 'react';
import { ComponentEnhancer, compose, setDisplayName ,withState } from 'recompose';
import primaryTheme from './primary';
import secondaryTheme from './secondary';

const themes = {
  primary: primaryTheme,
  secondary: secondaryTheme
};

export interface ThemeProviderProps {
  theme?: 'primary' | 'secondary';
  children?: React.SFC<any>;
};

export interface ThemeProviderDispatches {
  setTheme: (theme: 'primary' | 'secondary') => void;
};

export type EnhancedThemeProviderProps = ThemeProviderProps & ThemeProviderDispatches;

const ThemeProvider: React.SFC<EnhancedThemeProviderProps> = ({
  theme = 'primary',
  setTheme,
  children
}) => (
  <Provider theme={themes[theme]}>
    {children && children({ theme, setTheme })}
  </Provider>
);

const withTheme = compose(
  withState('theme', 'setTheme', ({ theme = 'primary' }: any) => theme)
);

const enhance = compose(
  setDisplayName('ThemeProvider'),
  withTheme
) as ComponentEnhancer<EnhancedThemeProviderProps, ThemeProviderProps>

export default enhance(ThemeProvider);
