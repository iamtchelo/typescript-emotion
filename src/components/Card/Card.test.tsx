import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import * as React from 'react';
import ThemeProvider from '../../themes';
import Card from './Card';

describe('<Card />', () => {
  const props = {
    image: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png',
    title: 'My card'
  }
  
  it('should render with props', () => {
    const tree = shallow(
      <ThemeProvider>
        <Card {...props} />
      </ThemeProvider>
    );

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});
