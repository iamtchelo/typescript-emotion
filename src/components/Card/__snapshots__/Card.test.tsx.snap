// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> should render with props 1`] = `
<ThemeProvider
  theme={
    Object {
      "button": Object {
        "background": "#FFFFFF",
        "backgroundActived": "#80a1c1",
        "backgroundHover": "#7593b0",
        "borderColor": "#80a1c1",
        "color": "#414141",
      },
      "card": Object {
        "titleColor": "#808080",
      },
      "colors": Object {
        "gray": "#414141",
        "grayDark": "#808080",
        "grayLight": "#E0E0E0",
        "primary": "#80a1c1",
        "primaryHover": "#7593b0",
        "white": "#FFFFFF",
      },
      "navbar": Object {
        "background": "#80a1c1",
        "titleColor": "#FFFFFF",
      },
    }
  }
>
  <Card
    image="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
    title="My card"
  />
</ThemeProvider>
`;
